### PR TITLE
Johnny/unsl2

### PR DESCRIPTION
**Pull request by JohnnyJumper**
This PR initializes the support for UNS Layer2 solution.

PR ended up very big. Sorry for that, I will try to explain below the key changes to make it easier to review. 

Key changes:

1. Updated the library configuration. Introduced UnsLocation with Layer1 and Layer2 fields for configurations of each layer respectively.
2. Moved old UNS class logic into new UNSLayer class.
3. Introduced internal AsyncResolver class to asynchronously call layer1 and layer2, wait for their results and then parse the errors or return unsafe [UNSLocation: ResultConsumer<Z>] for further parsing.
4. UNS class now uses AsyncResolver class to make the calls and in the case of batchOwners and tokensOwnedBy parses the results of both layers and combines them in one list excluding duplications.
5. Fixed issue with some methods in UNS class when they weren't throwing UnregisteredDomain when expected it. (methods: resolver and records). (Fixed the behavior with tests)
6. Fixed the race condition issue inside tokensOwnedBy by using a synchronous queue in the asynchronous code block. This effectively prevents several threads to try to update the same information at the same time and forces threads to wait for the previous one to finish with the updates. (thanks to @rommex for the advice)
7. Removed un-needed method "getNetwork" from top-level Resolution class as well as from NamingService protocol 
8. Redistribute tests across different files (files were getting too big to efficiently found anything in them)
9. Moved all checkErrors methods and testDomains into the new TestHelpers class as they are using across many files now.
10. Update and added new tests to cover layer2 resolution

Things I wish to change:
1. AsyncResolver class has 2 generic functions that are essentially the same. They differ only in function types that are being passed by and used. ( most of the functions expect 1 argument and some expect 2). If anyone knows a solution to combine those two in 1 please let me know. 
2. I blindly moved the test domains into TestHelper class without changing the variable names. In the future, I want to make it more clear and easier to use. 
3. Get rid of the semaphores in the code. 
 3.1 (Async await came to swift only in the latest versions which prevent me from using it everywhere without breaking changes for older wallets)
 3.2 removing semaphores now in favor of completion handlers will cause an avalanche of internal changes. (I will need to change the protocol for each function for each naming service to pass the completion handlers along with and write new ones for deeper levels (the ones that actually makes a network call to the blockchain providers) at this point it is easier to just rewrite the entire library from scratch using async-await and deprecate the current library in favor of the latest one. 
 If anyone knows a better solution for this please let me know as well.
 4. This is a silly one, but I wish to change the names for the arguments I need to pass along to AsyncResolver. arg and arg1 / arg2 are too generous. If anyone has better names please let me know :)

### CODE REVIEW

**Review by rommex — APPROVED** [review]

**Review by merenkoff — APPROVED** [review]
LGTM

File: Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+
+            guard result.count > 0 else {
+                throw ResolutionError.unregisteredDomain
+            }
+
+            if let domainName = ABIDecoder.decodeSingleType(type: .string, data: Data(hex: result[0].data)).value as? String {
+                return domainName
+            }
+            throw ResolutionError.unregisteredDomain
+        } catch APIError.decodingError {
```
Comment (DeRain) [inline Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift:300]: @JohnnyJumper What leads to decodingError? I was thinking that you just won't get event if domain name is not registered.

File: Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+                                    .callMethod(methodName: Self.tokenURIMethodName,
+                                                args: [tokenId]) {
+                let dict = result as? [String: Any]
+                if let val = dict?["0"] as? String {
+                    return val
+                }
+                throw ResolutionError.unregisteredDomain
+            }
+            throw ResolutionError.proxyReaderNonInitialized
+        } catch APIError.decodingError {
```
Comment (DeRain) [inline Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift:277]: @JohnnyJumper Could you describe the case when we're getting this type of error? What does it means?

File: Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+                if let val = dict?["0"] as? EthereumAddress {
+                    if !Utillities.isNotEmpty(val.address) {
+                        throw ResolutionError.unregisteredDomain
+                    }
+                    return val.address
+                }
+                throw ResolutionError.unregisteredDomain
+            }
+            throw ResolutionError.proxyReaderNonInitialized
+        } catch APIError.decodingError {
```
Comment (DeRain) [inline Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift:379]: Same as above

File: Tests/ResolutionTests/ResolutionUNSTests.swift [hunk 4]
```diff
@@ -0,0 +1,315 @@
+//
+//  UnsLayerTest.swift
```
Comment (DeRain) [inline Tests/ResolutionTests/ResolutionUNSTests.swift:2]: I think these tests are part of ResolutionTests.swift

**Review by DeRain — CHANGES_REQUESTED** [review]
@JohnnyJumper Looks great so far! Thank you for all of yours hard work with this PR - i know it has been tough one.

I've left a few minor comments to solve + it would be great to have tests for `Uns.swift` to test L1<>L2 resolution logic precisely: include edge cases errors, check that error return layer (a list of errors mentioned in Pivotal ticket, check different domains on different layers and on both).

File: Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+        let result = try record(domain: domain, key: key)
+        return result
+    }
+
+    // MARK: - Get Record
+    func record(domain: String, key: String) throws -> String {
+        let tokenId = super.namehash(domain: domain)
+        let result = try recordFromTokenId(tokenId: tokenId, key: key)
+        guard Utillities.isNotEmpty(result) else {
+            throw ResolutionError.recordNotFound
```
Comment (DeRain) [inline Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift:223]: @JohnnyJumper Please make sure that error include layer (check Pivotal ticket for reference of errors list)

**Review by DeRain — CHANGES_REQUESTED** [review]

**Comment by DeRain** [comment]
@JohnnyJumper Looks good. Could you clarify my questions about abicoder errors? Thanks.

File: Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift (hunk 1 above)
Comment (JohnnyJumper) [reply Sources/UnstoppableDomainsResolution/NamingServices/UNSLayer.swift]: I found out that our decoder is failing when blockchain returns execution reverted error. This was the case which was covered by our tests. I added the logic to parse the following but found a very strange behavior. Some providers return the execution reverted messages with some additional data and a different code.

Sometimes the code is 3 and sometimes the code is -32000. When the code is 3 provider also returns some binary data and the message looks like following: 
`"execution reverted: ERC721Metadata: URI query for nonexistent token"`
When blockchain provider returns code -32000 the message is simply `"execution reverted"`

This call will return the error code of 3:
```
URL: https://matic-testnet-archive-rpc.bwarelabs.com
Body: {
	"jsonrpc": "2.0",
	"id": "1",
	"method": "eth_call",
	"params": [
		{
			"data": "0xc87b56dd756e4e998dbffd803c21d23b06cd855cdc7a4b57706c95964a37e24b47c10fc9",
			"to": "0x332A8191905fA8E6eeA7350B5799F225B8ed30a9"
			
		},
		"latest"
		]
}
RESPONSE: {
    "jsonrpc": "2.0",
    "id": "1",
    "error": {
        "code": 3,
        "message": "execution reverted: ERC721Metadata: URI query for nonexistent token",
        "data": "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002f4552433732314d657461646174613a2055524920717565727920666f72206e6f6e6578697374656e7420746f6b656e0000000000000000000000000000000000"
    }
}
```

and this one will return error code of -32000
```
URL: https://rinkeby.infura.io/v3/3c25f57353234b1b853e9861050f4817
BODY: {
	"jsonrpc": "2.0",
	"id": "1",
	"method": "eth_call",
	"params": [
		{
			"data": "0xc87b56dd6d8b296e38dfd295f2f4feb9ef2721c48210b7d77c0a08867123d9bd5150cf47",
			"to": "0xE6729D224D00b3dd4FC731C4Ee3274E35Da06578"
			
		},
		"latest"
		]
}
RESPONSE: {
    "jsonrpc": "2.0",
    "id": "1",
    "error": {
        "code": -32000,
        "message": "execution reverted"
    }
}
```

**Review by DeRain — APPROVED** [review]
